File: backend/app/db/models/memory.py
"""Memories — the knowledge store / future-training data collected from the
platform (chats, calls, whispers, documents)."""
from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, Enum, Float, ForeignKey, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, UUIDMixin
from app.db.models.enums import MemorySourceType, MemoryStatus


class Memory(UUIDMixin, Base):
    __tablename__ = "memories"

    user_id: Mapped[str | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), index=True
    )
    content: Mapped[str] = mapped_column(Text)
    source_type: Mapped[MemorySourceType] = mapped_column(
        Enum(MemorySourceType, name="memory_source_type")
    )
    confidence: Mapped[float | None] = mapped_column(Float)
    status: Mapped[MemoryStatus] = mapped_column(
        Enum(MemoryStatus, name="memory_status"), default=MemoryStatus.PENDING
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

File: apps/web/src/lib/memory.ts
/** Memory (training-data review queue) API calls. Types mirror
 * backend/app/models/memories.py. */
import { apiFetch } from "./api-client";

export type MemorySourceType = "chat" | "call" | "whisper" | "document";
export type MemoryStatus = "pending" | "approved" | "rejected";

export interface MemoryRead {
  id: string;
  user_id: string | null;
  content: string;
  source_type: MemorySourceType;
  confidence: number | null;
  status: MemoryStatus;
  created_at: string;
}

export interface MemoryCreate {
  content: string;
  source_type: MemorySourceType;
  confidence?: number;
  user_id?: string;
}

export const SOURCE_LABEL: Record<MemorySourceType, string> = {
  chat: "Chat conversations",
  call: "Voice calls",
  whisper: "Operator whispers",
  document: "Documents",
};

/** Quality tier derived from the real confidence score (0-1) — there is no
 * separate "quality" field on the backend. */
export type DataQuality = "high" | "medium" | "low";

export function qualityFromConfidence(confidence: number | null): DataQuality {
  if (confidence == null) return "medium";
  if (confidence > 0.8) return "high";
  if (confidence > 0.5) return "medium";
  return "low";
}

export function listMemories(status?: MemoryStatus): Promise<MemoryRead[]> {
  const query = status ? `?status=${status}` : "";
  return apiFetch<MemoryRead[]>(`/api/memories${query}`);
}

export function createMemory(input: MemoryCreate): Promise<MemoryRead> {
  return apiFetch<MemoryRead>("/api/memories", { method: "POST", json: input });
}

export function updateMemoryStatus(
  id: string,
  status: MemoryStatus,
): Promise<MemoryRead> {
  return apiFetch<MemoryRead>(`/api/memories/${id}/status`, {
    method: "PATCH",
    json: { status },
  });
}
